File: Telegram/SourceFiles/mcp/mcp_server.cpp
// MCP Server implementation
//
// This file is part of Telegram Desktop MCP integration,
// the official desktop application for the Telegram messaging service.
//
// For license and copyright information please follow this link:
// https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL

#include "mcp_server.h"

#include <QtCore/QTimer>
#include <QtCore/QDebug>

namespace MCP {

Server::Server(QObject *parent)
	: QObject(parent) {
	initializeCapabilities();
	registerTools();
	registerResources();
	registerPrompts();
}

Server::~Server() {
	stop();
}

void Server::initializeCapabilities() {
	_serverInfo.capabilities = QJsonObject{
		{"tools", QJsonObject{{"listChanged", true}}},
		{"resources", QJsonObject{{"listChanged", true}}},
		{"prompts", QJsonObject{{"listChanged", true}}},
	};
}

void Server::registerTools() {
	_tools = {
		Tool{
			"list_chats",
			"Get a list of all Telegram chats (direct access to local database)",
			QJsonObject{
				{"type", "object"},
				{"properties", QJsonObject{}},
			}
		},
		Tool{
			"get_chat_info",
			"Get detailed information about a specific chat",
			QJsonObject{
				{"type", "object"},
				{"properties", QJsonObject{
					{"chat_id", QJsonObject{
						{"type", "integer"},
						{"description", "Telegram chat ID"}
					}}
				}},
				{"required", QJsonArray{"chat_id"}},
			}
		},
		Tool{
			"read_messages",
			"Read messages from local database (instant, no API calls!)",
			QJsonObject{
				{"type", "object"},
				{"properties", QJsonObject{
					{"chat_id", QJsonObject{
						{"type", "integer"},
						{"description", "Chat ID"}
					}},
					{"limit", QJsonObject{
						{"type", "integer"},
						{"description", "Number of messages"},
						{"default", 50}
					}}
				}},
				{"required", QJsonArray{"chat_id"}},
			}
		},
		Tool{
			"send_message",
			"Send a message to a chat",
			QJsonObject{
				{"type", "object"},
				{"properties", QJsonObject{
					{"chat_id", QJsonObject{
						{"type", "integer"},
						{"description", "Chat ID"}
					}},
					{"text", QJsonObject{
						{"type", "string"},
						{"description", "Message text"}
					}}
				}},
				{"required", QJsonArray{"chat_id", "text"}},
			}
		},
		Tool{
			"search_messages",
			"Search messages in local database (semantic search coming soon)",
			QJsonObject{
				{"type", "object"},
				{"properties", QJsonObject{
					{"query", QJsonObject{
						{"type", "string"},
						{"description", "Search query"}
					}},
					{"chat_id", QJsonObject{
						{"type", "integer"},
						{"description", "Optional: limit to specific chat"}
					}},
					{"limit", QJsonObject{
						{"type", "integer"},
						{"default", 50}
					}}
				}},
				{"required", QJsonArray{"query"}},
			}
		},
	};
}

void Server::registerResources() {
	_resources = {
		Resource{
			"telegram://chats",
			"All Chats",
			"List of all Telegram chats",
			"application/json"
		},
		Resource{
			"telegram://messages/{chat_id}",
			"Chat Messages",
			"Messages from a specific chat",
			"application/json"
		},
	};
}

void Server::registerPrompts() {
	_prompts = {
		Prompt{
			"summarize_chat",
			"Analyze and summarize recent messages in a chat",
			QJsonArray{
				QJsonObject{
					{"name", "chat_id"},
					{"description", "Chat ID to summarize"},
					{"required", true}
				},
				QJsonObject{
					{"name", "limit"},
					{"description", "Number of messages to analyze"},
					{"required", false}
				}
			}
		},
	};
}

bool Server::start(TransportType transport) {
	if (_initialized) {
		return true;
	}

	_transport = transport;

	switch (_transport) {
	case TransportType::Stdio:
		startStdioTransport();
		break;
	case TransportType::HTTP:
		startHttpTransport();
		break;
	default:
		qWarning() << "MCP: Unsupported transport type";
		return false;
	}

	_initialized = true;
	qInfo() << "MCP Server started (transport:"
		<< (_transport == TransportType::Stdio ? "stdio" : "http")
		<< ")";

	return true;
}

void Server::stop() {
	if (!_initialized) {
		return;
	}

	delete _stdin;
	delete _stdout;
	delete _httpServer;

	_stdin = nullptr;
	_stdout = nullptr;
	_httpServer = nullptr;

	_initialized = false;
	qInfo() << "MCP Server stopped";
}

void Server::startStdioTransport() {
	_stdin = new QTextStream(stdin);
	_stdout = new QTextStream(stdout);

	// Use a timer to poll stdin (Qt doesn't have native stdin notifications)
	QTimer *timer = new QTimer(this);
	connect(timer, &QTimer::timeout, this, &Server::handleStdioInput);
	timer->start(100); // Poll every 100ms
}

void Server::handleStdioInput() {
	if (!_stdin->atEnd()) {
		QString line = _stdin->readLine();

		if (line.isEmpty()) {
			return;
		}

		// Parse JSON-RPC request
		QJsonParseError error;
		QJsonDocument doc = QJsonDocument::fromJson(line.toUtf8(), &error);

		if (error.error != QJsonParseError::NoError) {
			qWarning() << "MCP: JSON parse error:" << error.errorString();
			return;
		}

		QJsonObject request = doc.object();
		QJsonObject response = handleRequest(request);

		// Write response to stdout
		*_stdout << QJsonDocument(response).toJson(QJsonDocument::Compact);
		*_stdout << "\n";
		_stdout->flush();
	}
}

void Server::startHttpTransport(int port) {
	// TODO: Implement HTTP transport with QHttpServer (Qt 6)
	// This would enable SSE notifications and web-based clients
	qInfo() << "MCP: HTTP transport on port" << port << "(not implemented yet)";
}

QJsonObject Server::handleRequest(const QJsonObject &request) {
	QString method = request["method"].toString();
	QJsonObject params = request["params"].toObject();
	QJsonValue id = request["id"];

	qDebug() << "MCP: Request" << method;

	// Dispatch to method handlers
	if (method == "initialize") {
		return successResponse(id, handleInitialize(params));
	} else if (method == "tools/list") {
		return successResponse(id, handleListTools(params));
	} else if (method == "tools/call") {
		return successResponse(id, handleCallTool(params));
	} else if (method == "resources/list") {
		return successResponse(id, handleListResources(params));
	} else if (method == "resources/read") {
		return successResponse(id, handleReadResource(params));
	} else if (method == "prompts/list") {
		return successResponse(id, handleListPrompts(params));
	} else if (method == "prompts/get") {
		return successResponse(id, handleGetPrompt(params));
	} else {
		return errorResponse(id, -32601, "Method not found: " + method);
	}
}

QJsonObject Server::handleInitialize(const QJsonObject &params) {
	Q_UNUSED(params);

	return QJsonObject{
		{"protocolVersion", "2024-11-05"},
		{"serverInfo", QJsonObject{
			{"name", _serverInfo.name},
			{"version", _serverInfo.version}
		}},
		{"capabilities", _serverInfo.capabilities},
	};
}

QJsonObject Server::handleListTools(const QJsonObject &params) {
	Q_UNUSED(params);

	QJsonArray tools;
	for (const auto &tool : _tools) {
		tools.append(QJsonObject{
			{"name", tool.name},
			{"description", tool.description},
			{"inputSchema", tool.inputSchema},
		});
	}

	return QJsonObject{{"tools", tools}};
}

QJsonObject Server::handleCallTool(const QJsonObject &params) {
	QString name = params["name"].toString();
	QJsonObject arguments = params["arguments"].toObject();

	qDebug() << "MCP: Calling tool" << name << "with args" << arguments;

	// Dispatch to tool implementations
	QJsonObject result;

	if (name == "list_chats") {
		result = toolListChats(arguments);
	} else if (name == "get_chat_info") {
		result = toolGetChatInfo(arguments);
	} else if (name == "read_messages") {
		result = toolReadMessages(arguments);
	} else if (name == "send_message") {
		result = toolSendMessage(arguments);
	} else if (name == "search_messages") {
		result = toolSearchMessages(arguments);
	} else {
		// Unknown tool
		return QJsonObject{
			{"isError", true},
			{"content", QJsonArray{
				QJsonObject{
					{"type", "text"},
					{"text", "Unknown tool: " + name}
				}
			}}
		};
	}

	// Format as MCP tool response
	QJsonObject contentItem;
	contentItem["type"] = "text";
	contentItem["text"] = QString::fromUtf8(QJsonDocument(result).toJson(QJsonDocument::Compact));

	QJsonArray contentArray;
	contentArray.append(contentItem);

	QJsonObject response;
	response["content"] = contentArray;
	return response;
}

QJsonObject Server::handleListResources(const QJsonObject &params) {
	Q_UNUSED(params);

	QJsonArray resources;
	for (const auto &resource : _resources) {
		resources.append(QJsonObject{
			{"uri", resource.uri},
			{"name", resource.name},
			{"description", resource.description},
			{"mimeType", resource.mimeType},
		});
	}

	return QJsonObject{{"resources", resources}};
}

QJsonObject Server::handleReadResource(const QJsonObject &params) {
	QString uri = params["uri"].toString();

	// TODO: Implement resource reading
	return QJsonObject{
		{"uri", uri},
		{"mimeType", "application/json"},
		{"text", "{}"},
	};
}

QJsonObject Server::handleListPrompts(const QJsonObject &params) {
	Q_UNUSED(params);

	QJsonArray prompts;
	for (const auto &prompt : _prompts) {
		prompts.append(QJsonObject{
			{"name", prompt.name},
			{"description", prompt.description},
			{"arguments", prompt.arguments},
		});
	}

	return QJsonObject{{"prompts", prompts}};
}

QJsonObject Server::handleGetPrompt(const QJsonObject &params) {
	QString name = params["name"].toString();

	// TODO: Generate prompt based on arguments
	return QJsonObject{
		{"description", "Prompt: " + name},
		{"messages", QJsonArray{}},
	};
}

// Tool implementations (stubs - will connect to real tdesktop data)

QJsonObject Server::toolListChats(const QJsonObject &args) {
	Q_UNUSED(args);

	// TODO: Access tdesktop's dialog list
	// Example: _session->data().chatsListFor(Data::Folder::kAll)

	return QJsonObject{
		{"chats", QJsonArray{}},
		{"source", "local_database"},
		{"note", "Will access tdesktop's local chat list"},
	};
}

QJsonObject Server::toolGetChatInfo(const QJsonObject &args) {
	qint64 chatId = args["chat_id"].toVariant().toLongLong();

	// TODO: Get chat from tdesktop
	// Example: _session->data().peer(peerFromUser(chatId))

	return QJsonObject{
		{"chat_id", chatId},
		{"source", "local_database"},
		{"note", "Will access tdesktop's chat data"},
	};
}

QJsonObject Server::toolReadMessages(const QJsonObject &args) {
	qint64 chatId = args["chat_id"].toVariant().toLongLong();
	int limit = args["limit"].toInt(50);

	// TODO: Access tdesktop's message history
	// Example: _session->data().history(peer)->messages()

	qDebug() << "MCP: read_messages chat_id=" << chatId << "limit=" << limit;

	return QJsonObject{
		{"chat_id", chatId},
		{"messages", QJsonArray{}},
		{"source", "local_database"},
		{"note", "Will access tdesktop's SQLite message cache - INSTANT!"},
	};
}

QJsonObject Server::toolSendMessage(const QJsonObject &args) {
	qint64 chatId = args["chat_id"].toVariant().toLongLong();
	QString text = args["text"].toString();

	// TODO: Send message via tdesktop
	// Example: _session->api().sendMessage(peer, text)

	return QJsonObject{
		{"chat_id", chatId},
		{"text", text},
		{"status", "pending"},
		{"note", "Will use tdesktop's message sending"},
	};
}

QJsonObject Server::toolSearchMessages(const QJsonObject &args) {
	QString query = args["query"].toString();

	// TODO: Search tdesktop's local database
	// Future: Semantic search with FAISS

	return QJsonObject{
		{"query", query},
		{"results", QJsonArray{}},
		{"source", "local_database"},
		{"note", "Will search tdesktop's local message cache"},
	};
}

QJsonObject Server::errorResponse(
	const QJsonValue &id,
	int code,
	const QString &message
) {
	return QJsonObject{
		{"jsonrpc", "2.0"},
		{"id", id},
		{"error", QJsonObject{
			{"code", code},
			{"message", message}
		}},
	};
}

QJsonObject Server::successResponse(
	const QJsonValue &id,
	const QJsonObject &result
) {
	return QJsonObject{
		{"jsonrpc", "2.0"},
		{"id", id},
		{"result", result},
	};
}

} // namespace MCP

File: python-bridge/mcp_server.py
"""
MCP Server implementation for Telegram.

Exposes Telegram functionality through the Model Context Protocol:
- Tools: Callable functions (list_chats, send_message, etc.)
- Resources: URI-based data access
- Prompts: Pre-configured AI prompts
"""

from typing import Any, Dict, List, Optional

import structlog
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import Tool, TextContent, Resource, Prompt

from .config import Config
from .telegram_client import TelegramClient

logger = structlog.get_logger()


class TelegramMCPServer:
    """
    MCP Server for Telegram integration.

    Exposes Telegram operations as MCP tools, resources, and prompts.
    """

    def __init__(self, config: Config):
        """
        Initialize MCP server.

        Args:
            config: Configuration object
        """
        self.config = config
        self.telegram = TelegramClient(config)

        # Create MCP server instance
        self.server = Server(config.mcp.server_name)

        # Register handlers
        self._register_tools()
        self._register_resources()
        self._register_prompts()

        logger.info(
            "mcp_server.initialized",
            server_name=config.mcp.server_name,
            version=config.mcp.server_version,
        )

    def _register_tools(self) -> None:
        """Register MCP tools."""

        @self.server.list_tools()
        async def list_tools() -> List[Tool]:
            """List available tools."""
            return [
                Tool(
                    name="list_chats",
                    description="Get a list of all accessible Telegram chats (groups, channels, private chats)",
                    inputSchema={
                        "type": "object",
                        "properties": {},
                    },
                ),
                Tool(
                    name="get_chat_info",
                    description="Get detailed information about a specific Telegram chat",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "chat_id": {
                                "type": "integer",
                                "description": "Telegram chat ID (can be positive or negative)",
                            },
                        },
                        "required": ["chat_id"],
                    },
                ),
                Tool(
                    name="read_messages",
                    description="Read recent messages from a Telegram chat (FULL history access via MTProto)",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "chat_id": {
                                "type": "integer",
                                "description": "Telegram chat ID",
                            },
                            "limit": {
                                "type": "integer",
                                "description": f"Number of messages to read (default: {self.config.limits.default_message_limit}, max: {self.config.limits.max_message_limit})",
                                "default": self.config.limits.default_message_limit,
                            },
                        },
                        "required": ["chat_id"],
                    },
                ),
                Tool(
                    name="send_message",
                    description="Send a message to a Telegram chat",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "chat_id": {
                                "type": "integer",
                                "description": "Telegram chat ID",
                            },
                            "text": {
                                "type": "string",
                                "description": "Message text (supports Markdown formatting)",
                            },
                        },
                        "required": ["chat_id", "text"],
                    },
                ),
                Tool(
                    name="get_user_info",
                    description="Get information about a Telegram user",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "user_id": {
                                "type": "integer",
                                "description": "Telegram user ID",
                            },
                        },
                        "required": ["user_id"],
                    },
                ),
                Tool(
                    name="get_chat_members",
                    description="Get list of members in a Telegram group or channel",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "chat_id": {
                                "type": "integer",
                                "description": "Telegram chat ID",
                            },
                        },
                        "required": ["chat_id"],
                    },
                ),
                Tool(
                    name="search_messages",
                    description="Search for messages in a Telegram chat",
                    inputSchema={
                        "type": "object",
                        "properties": {
                            "chat_id": {
                                "type": "integer",
                                "description": "Telegram chat ID",
                            },
                            "query": {
                                "type": "string",
                                "description": "Search query",
                            },
                            "limit": {
                                "type": "integer",
                                "description": "Maximum number of results (default: 50)",
                                "default": 50,
                            },
                        },
                        "required": ["chat_id", "query"],
                    },
                ),
            ]

        @self.server.call_tool()
        async def call_tool(name: str, arguments: Dict[str, Any]) -> List[TextContent]:
            """Execute a tool."""
            try:
                if name == "list_chats":
                    result = await self.telegram.get_chats()

                elif name == "get_chat_info":
                    chat_id = arguments["chat_id"]
                    result = await self.telegram.get_chat_info(chat_id)

                elif name == "read_messages":
                    chat_id = arguments["chat_id"]
                    limit = arguments.get("limit", self.config.limits.default_message_limit)
                    result = await self.telegram.read_messages(chat_id, limit)

                elif name == "send_message":
                    chat_id = arguments["chat_id"]
                    text = arguments["text"]
                    result = await self.telegram.send_message(chat_id, text)

                elif name == "get_user_info":
                    user_id = arguments["user_id"]
                    result = await self.telegram.get_user_info(user_id)

                elif name == "get_chat_members":
                    chat_id = arguments["chat_id"]
                    result = await self.telegram.get_chat_members(chat_id)

                elif name == "search_messages":
                    chat_id = arguments["chat_id"]
                    query = arguments["query"]
                    limit = arguments.get("limit", 50)
                    result = await self.telegram.search_messages(chat_id, query, limit)

                else:
                    raise ValueError(f"Unknown tool: {name}")

                return [TextContent(type="text", text=str(result))]

            except Exception as e:
                logger.error(
                    "mcp_server.tool_execution_failed",
                    tool=name,
                    error=str(e),
                )
                return [TextContent(type="text", text=f"Error: {str(e)}")]

    def _register_resources(self) -> None:
        """Register MCP resources."""

        @self.server.list_resources()
        async def list_resources() -> List[Resource]:
            """List available resources."""
            return [
                Resource(
                    uri="telegram://chats",
                    name="Telegram Chats",
                    description="List of all accessible Telegram chats",
                    mimeType="application/json",
                ),
            ]

        @self.server.read_resource()
        async def read_resource(uri: str) -> str:
            """Read a resource."""
            if uri == "telegram://chats":
                chats = await self.telegram.get_chats()
                return str(chats)

            raise ValueError(f"Unknown resource URI: {uri}")

    def _register_prompts(self) -> None:
        """Register MCP prompts."""

        @self.server.list_prompts()
        async def list_prompts() -> List[Prompt]:
            """List available prompts."""
            return [
                Prompt(
                    name="telegram_summary",
                    description="Analyze and summarize recent messages in a Telegram chat",
                    arguments=[
                        {
                            "name": "chat_id",
                            "description": "Telegram chat ID",
                            "required": True,
                        },
                        {
                            "name": "limit",
                            "description": "Number of messages to analyze",
                            "required": False,
                        },
                    ],
                ),
            ]

        @self.server.get_prompt()
        async def get_prompt(name: str, arguments: Dict[str, str]) -> str:
            """Get a prompt."""
            if name == "telegram_summary":
                chat_id = arguments["chat_id"]
                limit = int(arguments.get("limit", "20"))

                return f"""Analyze the last {limit} messages in Telegram chat {chat_id} and provide:

1. A brief summary of the conversation
2. Key topics discussed
3. Main participants and their roles
4. Any action items or decisions made
5. Overall sentiment of the conversation

Use the read_messages tool to fetch the messages first."""

            raise ValueError(f"Unknown prompt: {name}")

    async def start(self) -> None:
        """Start the MCP server and Telegram client."""
        # Start Telegram client
        await self.telegram.start()

        logger.info("mcp_server.started")

    async def stop(self) -> None:
        """Stop the MCP server and Telegram client."""
        # Stop Telegram client
        await self.telegram.stop()

        logger.info("mcp_server.stopped")

    async def run(self) -> None:
        """Run the MCP server with stdio transport."""
        try:
            # Start services
            await self.start()

            # Run server
            async with stdio_server() as (read_stream, write_stream):
                await self.server.run(
                    read_stream,
                    write_stream,
                    self.server.create_initialization_options(),
                )

        finally:
            # Cleanup
            await self.stop()
